elements/pl-big-o-input/big_o_utils.py

```diff
@@ -0,0 +1,252 @@
+
+    if sympy.limit(sym_sub, sympy.Symbol(variables[0]), sympy.oo) < sympy.sympify(0):
+        return (0, NEGATIVE_FEEDBACK)
+
+    L = sympy.limit(sym_true / sym_sub, sympy.Symbol(variables[0]), sympy.oo)
+
+    if L < sympy.sympify(0):
+        return (0, NEGATIVE_FEEDBACK)
+    elif L == sympy.oo:
+        return (0.25, TOO_LOOSE_FEEDBACK)
+    elif L == sympy.sympify(0):
+        return (0, INCORRECT_FEEDBACK)
+    elif L == sympy.sympify(1):
+        return (0.5, LOWER_ORDER_TERMS_FEEDBACK)
+
+    return (0.5, CONSTANT_FACTORS_FEEDBACK)
+
+
+# Start of generic utilities. Move these to prairielearn.py as-needed
+
+
+class DisplayType(Enum):
+    INLINE = "inline"
+    BLOCK = "block"
+
+
+def grade_question_parameterized(
+    data: pl.QuestionData,
+    question_name: str,
+    grade_function: Callable[[Any], Tuple[Union[bool, float], Optional[str]]],
+    weight: int = 1,
+) -> None:
+    """
+    Grade question question_name. grade_function should take in a single parameter
+    (which will be the submitted answer) and return a 2-tuple:
+        - The first element of the 2-tuple should either be:
+            - a boolean indicating whether the question should be marked correct
+            - a partial score between 0 and 1, inclusive
+        - The second element of the 2-tuple should either be:
+            - a string containing feedback
+            - None, if there is no feedback (usually this should only occur if the answer is correct)
+    """
+
+    # Create the data dictionary at first
+    data["partial_scores"][question_name] = {"score": 0.0, "weight": weight}
+
+    if question_name not in data["submitted_answers"]:
+        data["format_errors"][question_name] = "No answer was submitted"
+        return
+
+    submitted_answer = data["submitted_answers"][question_name]
+
+    # Run passed-in grading function
+    result, feedback_content = grade_function(submitted_answer)
+
+    # Try converting partial score
+    if isinstance(result, bool):
+        partial_score = 1.0 if result else 0.0
+    elif isinstance(result, (float, int)):
+        assert 0.0 <= result <= 1.0
+        partial_score = result
+    else:
+        assert_never(result)
+
+    # Set corresponding partial score and feedback
+    data["partial_scores"][question_name]["score"] = partial_score
+
+    if feedback_content:
+        data["partial_scores"][question_name]["feedback"] = feedback_content
+
+
+def determine_score_params(score: float) -> Tuple[str, Union[bool, float]]:
+    """Determine score params taken from data dict"""
+
+    if score >= 1:
+        return ("correct", True)
+    elif score > 0:
+        return ("partial", math.floor(score * 100))
+
+    return ("incorrect", True)
+
+
+EnumT = TypeVar("EnumT", bound=Enum)
+
+
+def get_enum_attrib(
+    enum_type: Type[EnumT],
+    element: lxml.html.HtmlElement,
+    name: str,
+    default: Optional[EnumT] = None,
+) -> EnumT:
+    """value = get_enum_attrib(enum_type, element, name, default)
+
```

Comment:
I think the signature should be immediately clear from the types?
```suggestion
    """
```

Reply:
This is just to match the docs for the other `get_*_attrib` functions.

Reply:
Tangentially related, @nwalters512 do you think the order of the arguments be changed for this function?  I think it feels a little different than the other getters since the element isn't the first argument.

Reply:
> This is just to match the docs for the other `get_*_attrib` functions.

Ah, gotcha, the other functions didn't show up in the diff. Especially with type info, I think we should remove that repeat of the function signatures from _all_ these functions, though I won't ask you to do that here.

I might consider putting `enum_type` after `name`?

Reply:
I think that's a reasonable course of action. There's a PR we can get in a bit later that adds type info for these functions (best done after typechecking is set for a few elements first), and we can change the docs there.

And sounds good, will change as a part of rebasing this.